### PR TITLE
[SPARK-29743][SQL] sample should set needCopyResult to true if its child is

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

`SampleExec` has a bug that it sets `needCopyResult` to false as long as the `withReplacement` parameter is false. This causes problems if its child needs to copy the result, e.g. a join.

### Why are the changes needed?

to fix a correctness issue

### Does this PR introduce any user-facing change?

Yes, the result will be corrected.

### How was this patch tested?

a new test

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @viirya @maropu

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala [hunk 1]
```diff
@@ -299,7 +299,9 @@ case class SampleExec(
     child.asInstanceOf[CodegenSupport].produce(ctx, this)
   }
 
-  override def needCopyResult: Boolean = withReplacement
+  override def needCopyResult: Boolean = {
+    child.asInstanceOf[CodegenSupport].needCopyResult || withReplacement
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: Good catch!

**Review by viirya — APPROVED** [review]

**Comment by gatorsmile** [comment]
This bug fix needs to be backported to all the active versions.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: Oh, yes. Indeed.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you so much, @cloud-fan !

**Comment by dongjoon-hyun** [comment]
I tested the new UT locally, but this PR might break some other `sample` UTs in `PySpark/SparkR`. Let's see the Jenkins result.

**Comment by dongjoon-hyun** [comment]
Merged to master/2.4. Thank you, @cloud-fan , @viirya , @gatorsmile .

**Comment by maropu** [comment]
oh.. I'm late. LGTM. Anyway, good catch!

**Pull request closed without merging**